doc/sphinx/proof-engine/ssreflect-proof-language.rst

```diff
@@ -2441,8 +2497,39 @@ intermediate statement.
 The have tactic.
 ````````````````
 
-.. tacn:: have : @term
-   :name: have
+.. tacn:: have @ssrhpats_wtransp {* @ssrbinder } @ssrhavefwd
+   :name: have (ssreflect)
+
+   .. insertprodn ssrhpats_wtransp ssrhavefwd
+
+   .. prodn::
+      ssrhpats_wtransp ::= {? @ssripats }
+      | {? @ssripats } @ {? @ssripats }
+      ssripats ::= {+ @i_item }
+      s_item ::= //=
+      | /=
+      | / @natural {| / {? @natural = } | = | /= }
+      | // {? @natural = }
+      | //
+      ssrdocc ::= %{ @clear_switch %}
+      | %{ {* @ident } %}
+      clear_switch ::= {| @natural | + | - } {* @natural }
```

Comment:
This is not the clear switch, I think the current parsing rule "mix" `clear_switch` and `occ_switch` in order to please camlp5, but the the code only accepts one or the other.
The switch for occurrence selection looks like `{+ 1 3 44}` while the one for clearing looks like `{H1 H2 H3}`.

Reply:
I guessed `ssrocc` in the mlg was really `clear_switch`, so I renamed it in the doc.  I couldn't guess what `occ_switch` in the old doc corresponds to in the mlg.  Maybe `ssrrwocc`?:


![image](https://user-images.githubusercontent.com/1253341/97258767-66cbbb00-17d6-11eb-8443-dd5583e2c0e3.png)

Reply:
Hum, I'm not sure I understand. Let's restart.
- `clear_switch` is currently misnamed, since it only encompasses numbers, which are to designate occurrences. Probably `occurrence_selection` is more appropriate. In between `{}` one can put `+` or `-` followed by 0 or more numbers, or no +/- but one or more numbers.
- `clear_switch` should be the name of the production containing zero o more idents between {}. `{}` is a clear switch but not an occurrence selection.
- `r_prefix` is misnamed, here it denotes a rewrite "item" in the sense that `rewrite item1 ... itemn` is the general syntax for rewrite. Each item can be:
  - `dir {occ} [pat]/term` an unfolding, with optional occurrence selection, optional pattern between square brackets, optional direction (nothing or `-`)
  - `{H1 .. HN}` a clear switch
  - `dir mult {occ} [pat] term` a rewrite rule, all optional but term, mult is ! or ? optionally prefixed by a positive number as in `2?` or `-1!` (- is the direction)
  - `/=` or `//` or `//=` a simplification

Reply:
I made some changes, but let's take it step by step.  Generally, we use non-equivalent grammar in the doc to hide unsupported combinations (e.g. semantic restrictions, useless, perhaps Camlp5 artifacts).  If we find simple bugs in the mlg grammar, we try to fix those.  But otherwise, we want the doc to be faithful to the actual mlg syntax.

- What is `dir`?
- `- {? mult }` and `mult` should be combined in these two productions in the obvious way? `-` doesn't mean a negative number here?

![image](https://user-images.githubusercontent.com/1253341/98174101-25f43600-1ea9-11eb-84d3-39cb574dfd2c.png)

- These two productions should be combined into one, with the part before `ssrpattern_squarep` a reference to `ssrrwocc` (to be renamed `occ_or_clear`)?  The differences in what's optional (ident+ vs ident* and optional/required clear_switch) are parser artifacts?

![image](https://user-images.githubusercontent.com/1253341/98175052-be3eea80-1eaa-11eb-89d8-15af28839114.png)

Reply:
`dir` stands for "direction". In vanilla Coq you have `->` for left to right and `<-` for right to left. `->` can be omitted, its the default. In ssr you have "nothing" for `->` and `-` for `<-`. So `dir` is wither empty or `-`.

Inside `{}` where you can specify occurrence numbers, `{- 1 7}` means "all but 1 and 7", not at all the same `-`.

`mult` is for multiplicity. Nothing means once. `!` mean 1 or more, `?` means zero o more. `10!` mean "one of more, up to 10`, `10?` is zero o more up to 10. I was hinting at the fact that `-10!` means "right-to-left, one or more up to 10", and not "one or more, up to -10".